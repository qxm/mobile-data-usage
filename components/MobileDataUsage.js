import React, { PropTypes, Component } from 'react';
import {
  ListView,
  StyleSheet,
  View,
  Modal,
  TouchableOpacity,
  WebView,
  RefreshControl,
  ActivityIndicator
} from 'react-native';
import DataUsageItem from './DataUsageItem';
import SmallText from './SmallText';
import * as globalStyles from '../styles/global';

export default class MobileDataUsage extends Component {

  constructor(props) {
    super(props);
    this.ds = new ListView.DataSource({
      rowHasChanged: (row1, row2) => row1._id !== row2._id
    });
    this.state = {
      dataSource: this.ds.cloneWithRows(props.dataUsage),
      initialLoading: true,
      modalVisible: false,
      refreshing: false
    };

    this.renderRow = this.renderRow.bind(this);
    this.onModalClose = this.onModalClose.bind(this);
    this.onModalOpen = this.onModalOpen.bind(this);
    this.refresh = this.refresh.bind(this);
  }

  componentWillMount() {
    this.refresh();
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      dataSource: this.state.dataSource.cloneWithRows(nextProps.dataUsage),
      initialLoading: false
    });
  }

  onModalClose() {
    this.setState({
      modalVisible: false,
      modalUrl: undefined
    });
  }

  onModalOpen(url) {
    this.setState({
      modalVisible: true,
      modalUrl: url
    });
  }

  refresh() {
    if (this.props.loadData) {
      this.props.loadData();
    }
  }

  renderModal() {
    return (
      <Modal
        animationType="slide"
        visible={this.state.modalVisible}
        onRequestClose={this.onModalClose}
      >
        <View style={styles.modalContent}>
          <TouchableOpacity
            onPress={this.onModalClose}
            style={styles.closeButton}
          >
            <SmallText>Close</SmallText>
          </TouchableOpacity>
          <WebView
            scalesPageToFit
            source={{ uri: this.state.modalUrl }}
          />
        </View>
      </Modal>
    );
  }

  renderRow(rowData, ...rest) {
    const index = parseInt(rest[1], 10);
    return (
      <DataUsageItem
        onPress={() => this.onModalOpen('https://www.google.com')}
        style={styles.dataUsageItem}
        index={index}
        data={[{...rowData}]}
      />
    );
  }

  render() {
    const {
      listStyles = globalStyles.COMMON_STYLES.pageContainer,
      showLoadingSpinner
    } = this.props;
    const { initialLoading, refreshing, dataSource } = this.state;

    return (
      (initialLoading && showLoadingSpinner
        ? (
          <View style={[listStyles, styles.loadingContainer]}>
            <ActivityIndicator
              animating
              size="small"
              {...this.props}
            />
          </View>
        ) : (
          <View style={styles.container}>
            <ListView
              refreshControl={
                <RefreshControl
                  refreshing={refreshing}
                  onRefresh={this.refresh}
                />
              }
              enableEmptySections
              dataSource={dataSource}
              renderRow={this.renderRow}
              style={listStyles}
            />
            {this.renderModal()}
          </View>
        )
      )
    );
  }
}

MobileDataUsage.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  listStyles: View.propTypes.style,
  loadData: PropTypes.func,
  showLoadingSpinner: PropTypes.bool
};

MobileDataUsage.defaultProps = {
  showLoadingSpinner: true
};

const styles = StyleSheet.create({
  dataUsageItem: {
    marginBottom: 20
  },
  container: {
    flex: 1
  },
  loadingContainer: {
    alignItems: 'center',
    justifyContent: 'center'
  },
  modalContent: {
    flex: 1,
    justifyContent: 'center',
    paddingTop: 20,
    backgroundColor: globalStyles.BG_COLOR
  },
  closeButton: {
    paddingVertical: 5,
    paddingHorizontal: 10,
    flexDirection: 'row'
  }
});
