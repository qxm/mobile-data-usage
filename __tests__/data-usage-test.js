import React from 'react';
import MobileDataUsage from '../components/MobileDataUsage';

import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const tree = renderer.create(<MobileDataUsage />).toJSON();
  expect(tree).toMatchSnapshot();
});
