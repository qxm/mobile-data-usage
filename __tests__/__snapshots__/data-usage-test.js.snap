// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#343336",
        "flex": 1,
        "marginBottom": 48,
        "marginHorizontal": 0,
        "marginTop": 0,
        "paddingHorizontal": 10,
        "paddingTop": 20,
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    showLoadingSpinner={true}
    size="small"
  />
</View>
`;
